keeper/src/main/scala/zio/keeper/membership/swim/Nodes.scala

```diff
@@ -5,121 +5,89 @@ import zio.clock.Clock
 import zio.logging._
 import zio.stm.TMap
 import zio.stream._
 
-/**
- * Nodes maintains state of the cluster.
- *
- * @param nodeStates - states
- * @param roundRobinOffset - offset for round-robin
- */
-final class Nodes(
-  nodeStates: TMap[NodeAddress, NodeState],
-  roundRobinOffset: Ref[Int],
-  eventsQueue: Queue[MembershipEvent],
-  internalEventsQueue: Queue[NodeStateChanged]
-) {
-
-  def addNode(node: NodeAddress) =
-    nodeStates
-      .put(node, NodeState.Init)
-      .whenM(nodeStates.contains(node).map(!_))
-      .commit
-      .unit
-
-  /**
-   * Changes node state and issue membership event.
-   * @param id - member id
-   * @param newState - new state
-   */
-  def changeNodeState(id: NodeAddress, newState: NodeState): ZIO[Logging, Error, Unit] =
-    nodeState(id)
-      .flatMap { prev =>
-        ZIO.when(prev != newState) {
-          log.info(s"changing node[$id] status from: [$prev] to: [$newState]") *>
-            nodeStates
-              .put(id, newState)
-              .commit
-              .tap(
-                _ => {
-                  ZIO.whenCase(newState) {
-                    case NodeState.Healthy if prev == NodeState.Init => eventsQueue.offer(Join(id))
-                    case NodeState.Dead | NodeState.Left             => eventsQueue.offer(Leave(id))
-                  } *> internalEventsQueue.offer(NodeStateChanged(id, prev, newState)).unit
-                }
-              )
-        }
-      }
-
-  /**
-   * close connection and remove Node from cluster.
-   * @param id node id
-   */
-  def disconnect(id: NodeAddress): IO[Error, Unit] =
-    nodeStates.delete(id).commit
+object Nodes {
 
   /**
-   *  Stream of Membership Events
+   * Nodes maintains state of the cluster.
    */
-  def events: Stream[Nothing, MembershipEvent] =
-    ZStream.fromQueue(eventsQueue)
+  trait Service {
+    def addNode(node: NodeAddress): UIO[Unit]
+
+    /**
+     * Changes node state and issue membership event.
+     * @param id - member id
+     * @param newState - new state
+     */
+    def changeNodeState(id: NodeAddress, newState: NodeState): IO[Error, Unit]
+
+    /**
+     * close connection and remove Node from cluster.
+     * @param id node id
+     */
+    def disconnect(id: NodeAddress): IO[Error, Unit]
+
+    /**
+     *  Stream of Membership Events
+     */
+    def events: Stream[Nothing, MembershipEvent]
+
+    /**
+     *  Stream of Membership Events for internal purpose.
+     *  This exists only because I cannot find the way to duplicates events from one queue
+     */
+    val internalEvents: Stream[Nothing, NodeStateChanged]
+
+    /**
+     * Returns next Healthy node.
+     */
+    val next: UIO[Option[NodeAddress]]
+
+    /**
+     * Node state for given NodeId.
+     */
+    def nodeState(id: NodeAddress): IO[Error, NodeState]
+
+    def numberOfNodes: UIO[Int]
+
+    /**
+     * Lists members that are in healthy state.
+     */
+    def healthyNodes: UIO[List[(NodeAddress, NodeState)]]
```

Comment:
Does it make sense to change to `Chunk`? Long-term goal in core is to migrate wherever possible. No need to do this now, just asking.

Reply:
do you mean migrate ```List``` to ```Chunk```?

Reply:
Yeah, I think we should start defaulting to Chunk.

Reply:
let me try it.

Reply:
let do this when `TMap` will have `toChunk` method.

Reply:
It's already there as of RC19.

Reply:
let do this in new PR